aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md

```diff
@@ -0,0 +1,202 @@
+
+[!INCLUDE[](~/includes/not-latest-version.md)]
+
+-->
+
+This article describes browser developer tools diagnostic tools in Blazor WebAssembly apps.
+
+## Prerequisite
+
+Install the [.NET WebAssembly build tools](xref:blazor/tooling/webassembly#net-webassembly-build-tools):
+
+```dotnetcli
+dotnet workload install wasm-tools
+```
+
+## Browser developer tools
+
+App code can be manually profiled using the diagnostic profiler in a browser's developer tools console.
+
+Built-in diagnostic counters are available to track:
+
+* [Ahead-of-time (AOT) compilation](xref:blazor/tooling/webassembly#ahead-of-time-aot-compilation)
+* Code interpolation
+* Call specification (":::no-loc text="callspec":::", sequence and timing of function calls) and instrumentation
+
+The MSBuild properties in the following table enable profiler integration.
+
+Property | Default | Set value to&hellip; | Description
+--- | :---: | :---: | ---
+`<WasmProfilers>` | No value | `browser` | Mono profilers to use. Potential values are "`browser`" and "`log`". To use both, separate the values with a semicolon. The `browser` profiler enables integration with the browser's developer tools profiler.
+`<RunAOTCompilation>`| `false` | `true` | Controls AOT compilation.
+`<RunAOTCompilationAfterBuild>` | `false` | `true` | Controls AOT compilation after build. By default, it's run only for publish.
+`<WasmNativeStrip>` | `true` | `false` | Controls stripping the native executable.
+`<WasmNativeDebugSymbols>` | `true` | `true` | Controls building with native debug symbols.
+`<WasmBuildNative>` | `false` | `true` | Controls building the native executable.
+
+Enabling profilers has negative size and performance impact, so don't publish an app for production with profilers enabled. In the following example, a condition is set on a property group section that only enables profiling when the app is built with `/p:BlazorSampleProfilingEnabled=true` (.NET CLI) or `<BlazorSampleProfilingEnabled>true</BlazorSampleProfilingEnabled>` in a Visual Studio publish profile, where "`BlazorSampleProfilingEnabled`" is a custom symbol name that you choose and doesn't conflict with other symbol names.
+
+In the app's project file (`.csproj`):
+
+```xml
+<PropertyGroup Condition="'$(BlazorSampleProfilingEnabled)' == 'true'">
+  <WasmProfilers>browser;</WasmProfilers>
+  <WasmNativeStrip>false</WasmNativeStrip>
+  <WasmNativeDebugSymbols>true</WasmNativeDebugSymbols>
+</PropertyGroup>
+```
+
+Alternatively, enable features when the app is built with the .NET CLI. The following options passed to the `dotnet build` command mirror the preceding MS Build property configuration:
+
+```dotnetcli
+/p:WasmProfilers=browser /p:WasmNativeStrip=false /p:WasmNativeDebugSymbols=true
+```
+
+Setting WebAssembly profilers with `<WasmProfilers>browser;</WasmProfilers>` doesn't require AOT (`<RunAOTCompilation>`/`<RunAOTCompilationAfterBuild>` set to `false` or removed from the preceding property group).
```

Comment:
"doesn't require AOT" is good, I would stop there.

Reply:
```suggestion
Setting WebAssembly profilers with `<WasmProfilers>browser;</WasmProfilers>` doesn't require AOT.
```